### PR TITLE
Added a warning for when function returns a value but has return void

### PR DESCRIPTION
**Pull request by thk123**
Also included a simple regression that should produce this warning

### CODE REVIEW

**Comment by thk123** [comment]
@cristina-david I think I should assign this to you, but I can't work out how to change the assignee. I suppose it is a permissions thing? 

For context, I thought I'd got stuck trying to get the interpreter to give me the return variable, but actually I was being silly and had the return type as void. When Chris spotted this after I asked for help, he suggested I added a warning to the C to GOTO compiler to catch this mistake.

File: src/ansi-c/c_typecheck_code.cpp [hunk 1]
```diff
@@ -911,7 +911,16 @@ void c_typecheck_baset::typecheck_return(codet &code)
     {
       // gcc doesn't actually complain, it just warns!
       if(follow(code.op0().type()).id()!=ID_empty)
+      {
+        warning() << "Function " << code.source_location().get_function();
+        warning() << " has return void but a return statement returning ";
+        warning() << follow(code.op0().type()).id();
```
Comment (kroening) [inline src/ansi-c/c_typecheck_code.cpp]: Use to_string here, to print the usual ANSI-C syntax

File: src/ansi-c/c_typecheck_code.cpp [hunk 2]
```diff
@@ -911,7 +911,16 @@ void c_typecheck_baset::typecheck_return(codet &code)
...
+        warning() << " was found on line " << code.source_location().get_line();
```
Comment (kroening) [inline src/ansi-c/c_typecheck_code.cpp]: Use warning().set_source_location instead.

File: src/ansi-c/c_typecheck_code.cpp [hunk 3]
```diff
@@ -911,7 +911,15 @@ void c_typecheck_baset::typecheck_return(codet &code)
     {
       // gcc doesn't actually complain, it just warns!
       if(follow(code.op0().type()).id()!=ID_empty)
+      {
+        warning().source_location = code.source_location();
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_code.cpp]: nit picking: don't put spaces around assignment operator

File: regression/ansi-c/function_return1/test.desc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--verbosity 2
+^file main.c line . function fun: Function has return void but a return statement returning signed int$
```
Comment (tautschnig) [inline regression/ansi-c/function_return1/test.desc]: I would suggest to insist the warning checks specifically for line 3

File: src/ansi-c/c_typecheck_code.cpp [hunk 5]
```diff
@@ -911,7 +911,15 @@ void c_typecheck_baset::typecheck_return(codet &code)
...
+
+        warning() << "Function has return void but a return statement returning ";
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_code.cpp]: We seem to be starting all warnings with a lowercase character (i.e., use "function")

Also, that line is rather long, consider breaking it earlier.

**Comment by tautschnig** [comment]
@thk123 Many thanks for carefully addressing all the feedback! It might be best to now squash all three commits into one.
@mgudemann I don't think your regression test actually relates to this issue?

**Comment by mgudemann** [comment]
@tautschnig You are right, I mentioned an internal issue `#301` in the commit message of another PR and github automagically referenced this here. This was definitely not intended.

**Comment by kroening** [comment]
This seems ready to get merged once it's squashed.

**Pull request merged**